File: services/nodejs-microservices/product-service/src/infrastructure/web/dto.ts
import { Product, Category, ProductReview } from '../../domain/models/Product';

export interface CreateProductRequest {
  name: string;
  description?: string;
  price: number;
  categoryId?: string;
  imageUrl?: string;
}

export interface UpdateProductRequest {
  name?: string;
  description?: string;
  price?: number;
  categoryId?: string;
  imageUrl?: string;
  status?: string;
}

export interface CreateCategoryRequest {
  name: string;
  description?: string;
  parentId?: string;
}

export interface CreateReviewRequest {
  userId: string;
  rating: number;
  comment?: string;
}

export interface ProductResponse {
  id: string;
  name: string;
  description?: string;
  price: number;
  categoryId?: string;
  imageUrl?: string;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export function toProductResponse(product: Product): ProductResponse {
  return {
    id: product.id,
    name: product.name,
    description: product.description,
    price: product.price,
    categoryId: product.categoryId,
    imageUrl: product.imageUrl,
    status: product.status,
    createdAt: product.createdAt.toISOString(),
    updatedAt: product.updatedAt.toISOString(),
  };
}

export function toCategoryResponse(category: Category) {
  return {
    id: category.id,
    name: category.name,
    description: category.description,
    parentId: category.parentId,
    createdAt: category.createdAt.toISOString(),
  };
}

export function toReviewResponse(review: ProductReview) {
  return {
    id: review.id,
    productId: review.productId,
    userId: review.userId,
    rating: review.rating,
    comment: review.comment,
    createdAt: review.createdAt.toISOString(),
  };
}

File: services/python-microservices/product-service/app/infrastructure/web/dto.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from decimal import Decimal


class CreateProductRequest(BaseModel):
    name: str
    description: Optional[str] = None
    price: Decimal
    category_id: Optional[str] = None
    image_url: Optional[str] = None


class UpdateProductRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[Decimal] = None
    category_id: Optional[str] = None
    image_url: Optional[str] = None
    status: Optional[str] = None


class ProductResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    price: Decimal
    category_id: Optional[str] = None
    image_url: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime


class CreateCategoryRequest(BaseModel):
    name: str
    description: Optional[str] = None
    parent_id: Optional[str] = None


class CategoryResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    parent_id: Optional[str] = None
    created_at: datetime


class CreateReviewRequest(BaseModel):
    user_id: str
    rating: int
    comment: Optional[str] = None


class ReviewResponse(BaseModel):
    id: str
    product_id: str
    user_id: str
    rating: int
    comment: Optional[str] = None
    created_at: datetime
